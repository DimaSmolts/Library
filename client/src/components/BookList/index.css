.book-list {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
  grid-gap: 1em;
  margin: 10px 0;
}
