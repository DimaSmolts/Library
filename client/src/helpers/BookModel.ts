export class Book {
  id: number;
  name: string;
  author: string;
  genres: Array<string>;
  description: string;
  rate: number;
  price: number;
  image: string;
}
