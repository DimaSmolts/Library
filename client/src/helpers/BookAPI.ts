import axios from 'axios';
import { Book } from './BookModel'

const baseURL = process.env.REACT_APP_API_URL || 'http://localhost:3000';

const instance = axios.create({
  baseURL: baseURL + '/api',
})

export const bookAPI = {
  async getBooks() {
    return await instance.get<Array<Book>>('/books');
  },
  async getBook(id: number) {
    return await instance.get<Book>('/books/' + id);
  }
}
