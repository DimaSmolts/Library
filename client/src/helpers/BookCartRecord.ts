import { Book } from './BookModel'

export interface BookCartRecord {
  book: Book,
  quantity: number,
}
