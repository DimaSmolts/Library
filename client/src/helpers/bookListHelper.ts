import { Book } from './BookModel'

export function filterBookList(books: Array<Book>, query: string): Array<Book> {
  const temp = books.filter(book => {
    return book.name
      .trim()
      .toLowerCase()
      .includes(query.trim().toLowerCase())
  })
  return temp;
}
