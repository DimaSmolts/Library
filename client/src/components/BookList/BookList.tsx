import React from 'react';
import { Book } from '../../helpers/BookModel'
import './index.css'

interface BookDListProps {
  books: Array<Book>,
}

export default class BookList extends React.Component<BookDListProps> {
  render() {
    // TODO: map book array to BookCards here
    return <span> hey </span>
  }
}
