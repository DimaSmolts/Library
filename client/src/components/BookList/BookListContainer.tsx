import React from 'react';
import BookList from './BookList'
import Message, { MessageTypes, MessageTextSizes } from '../Message/Message'
import { getBooksThunk, cleanUpThunk } from '../../store/thunks/bookListThunks'
import { connect, ConnectedProps } from 'react-redux';
import { AppState } from '../../store/rootReducer';

class BooksListContainer extends React.Component<PropsFromRedux> {
  componentDidMount() {
    this.props.getBooks();
  }

  componentWillUnmount() {
    this.props.cleanUp();
  }

  render() {
    if (this.props.error)
      throw new Error(this.props.error.toString());

    return <>
      {
        this.props.areFetched ?
          <BookList books={this.props.books} />
          :
          <Message type={MessageTypes.msgInfo} size={MessageTextSizes.msgBig} text='Loading...' />
      }
    </>
  }
}

const mapState = (state: AppState) => ({
  books: state.bookListReducer.books,
  areFetched: state.bookListReducer.areFetched,
  error: state.bookListReducer.areFetched
})


const mapDispatch = {
  getBooks: () => getBooksThunk(),
  cleanUp: () => cleanUpThunk(),
}

const connector = connect(mapState, mapDispatch);
type PropsFromRedux = ConnectedProps<typeof connector>

export default connector(BooksListContainer)
