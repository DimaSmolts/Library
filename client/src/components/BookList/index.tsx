import React from 'react'
import BookList from './BookList'
import Message, { MessageTypes, MessageTextSizes } from '../Message/Message'
import { connect, ConnectedProps } from 'react-redux'
import { AppState } from '../../store/rootReducer'

class BooksListContainer extends React.Component<PropsFromRedux> {
  render() {
    if (this.props.error)
      throw new Error(this.props.error.toString())
    if (this.props.notFound)
      return <Message type={MessageTypes.msgNotFound} size={MessageTextSizes.msgBig} text='Not Found' />

    return <>
      {
        this.props.areBooksFetched ?
          <BookList books={this.props.books} />
          :
          <Message type={MessageTypes.msgInfo} size={MessageTextSizes.msgBig} text='Loading...' />
      }
    </>
  }
}

const mapState = (state: AppState) => ({
  books: state.bookListReducer.books,
  areBooksFetched: state.bookListReducer.areFetched,
  error: state.bookListReducer.error,
  notFound: state.bookListReducer.notFound
})

const connector = connect(mapState, null)
type PropsFromRedux = ConnectedProps<typeof connector>

export default connector(BooksListContainer)
