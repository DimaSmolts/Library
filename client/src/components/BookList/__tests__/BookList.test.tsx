import React from 'react'
import { shallow } from 'enzyme'

import BookList from '../BookList'
import { Book } from '../../../helpers/BookModel'

const books = [{}, {}, {}, {}] as Array<Book>

test('snapshot: <BookList /> renders correctly', () => {
  const tree = shallow(<BookList books={books} />)

  expect(tree).toMatchSnapshot()
})

it('should render all book cards', () => {
  const wrapper = shallow(<BookList books={books} />)

  expect(wrapper.find('div').children().length).toBe(books.length)
})
