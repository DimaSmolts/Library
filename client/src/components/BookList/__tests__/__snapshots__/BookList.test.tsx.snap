// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot: <BookList /> renders correctly 1`] = `
<div
  className="book-list"
>
  <BookCard
    book={Object {}}
  />
  <BookCard
    book={Object {}}
  />
  <BookCard
    book={Object {}}
  />
  <BookCard
    book={Object {}}
  />
</div>
`;
